src/main/java/com/google/sps/servlets/BooksServlet.java

```diff
@@ -0,0 +1,142 @@
+  }
+
+  private String encodeBookMapAsJson() {
+    Gson gson = new Gson();
+    Map<String, Object> books = new HashMap<>();
+    books.put("books", booksMap);
+    String json = gson.toJson(books); 
+    return json;
+  }
+
+ /**
+  * Adds Book object to books from JSON results of Google Books query.
+  */
+  private void addBooksToMap(String jsonString, String keyTerm) {
+    JsonObject responseObject = new JsonParser().parse(jsonString).getAsJsonObject();
+    // Check to see if json has enough entries.
+    int numResults = responseObject.get("totalItems").getAsInt();
+    numResults = Math.min(numResults, NUM_BOOKS_PER_KEYWORD);
+    if (numResults <= 0) {
+      System.err.println("Error: no results found for query");
+      return;
+    }
+    JsonArray booksJsonArray = responseObject.getAsJsonArray("items");
+    List<Book> books = new ArrayList<>();
+    for (int i = 0; i < numResults; i++) {
+      JsonObject bookJson = booksJsonArray.get(i).getAsJsonObject().getAsJsonObject("volumeInfo"); 
+      String title = bookJson.get("title").getAsString();
+      String link = bookJson.get("infoLink").getAsString();
+      String image = bookJson.getAsJsonObject("imageLinks").get("thumbnail").getAsString();
+      String description = bookJson.get("description").getAsString();
+      String writer = formatWriters(bookJson.getAsJsonArray("authors"));
+      books.add(new Book.Builder(title, link).withImage(image).withDescription(description).withWriter(writer).build());
+    }
+    booksMap.put(keyTerm, books);
+  }
+
+ /**
+  * Reads in JSON from an open URL.
+  * @return JSON in String format.
+  */
+  private String readJsonFile(URL url) throws IOException {
+    Scanner jsonReader = new Scanner(url.openStream());
+    String data = "";
+    while(jsonReader.hasNext()) {
+        data += jsonReader.nextLine();
+    }
+    jsonReader.close();
+    return data;
+  }
+
+ /**
+  * Formats writers into string from JSON array.
+  * @return writers in String fomat.
+  */
+  private String formatWriters(JsonArray writersArray) {
+    String returnString = "";
+    int length = writersArray.size();
+    for (int i = 0; i < length - 1; i++){
+      returnString += writersArray.get(i).getAsString() + ", ";
+    }
+    returnString += writersArray.get(length - 1).getAsString();
+    return returnString;
+  }
+
+ /**
+  * Adds books objects for a key term.
+  */
+  private void addBooksForTerm(String keyTerm) {
+    String path = "https://www.googleapis.com/books/v1/volumes?";
+    String queryParam = "q=" + encodeTerm(keyTerm); 
+    String queryPath = path + queryParam;
+    try {
+      URL url = new URL(queryPath);
+      HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+      connection.setRequestMethod("GET");
+      connection.connect();
+      int responseCode = connection.getResponseCode();
+      if (responseCode == 200) {
+        String jsonString = readJsonFile(url);
+        addBooksToMap(jsonString, keyTerm);
```

Comment:
This one line is very different from all of the rest of the code in this function and is hard to find. You can factor it out by having this function just return the json and then passing that json into addBooksToMap in the parent function